crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -3294,88 +3294,104 @@ impl<'db> TypeInferenceBuilder<'db> {
         todo_type!("generic `typing.Awaitable` type")
     }
 
-    /// Look up a name reference that isn't bound in the local scope.
+    /// Look up a name reference that doesn't have any visible bindings from the local scope.
+    ///
+    /// Note that this does not mean that there are no bindings for the name in the current scope!
+    /// There *could* be bindings for the name in the local scope, and this function would still
+    /// be called if none of those bindings are visible from where we currently are in the AST
+    /// when this method is called.
```

Comment:
@carljm, I'd appreciate it if you could confirm that my understanding here is correct...!

Reply:
My only quibble is that I think it'd be clearer to say "where we currently are in control flow" rather than "where we currently are in the AST", but this looks right.

FWIW I'm not sure if this is ideal, though -- I think the split between what we do in `infer_name_load` and `lookup_name` doesn't make a ton of sense, and the code might be clearer if we moved all the logic into `lookup_name` such that it is called for every name load? But not 100% sure without trying it.

Reply:
> FWIW I'm not sure if this is ideal, though -- I think the split between what we do in `infer_name_load` and `lookup_name` doesn't make a ton of sense, and the code might be clearer if we moved all the logic into `lookup_name` such that it is called for every name load? But not 100% sure without trying it.

Excellent idea. I tried it, and I think it helps quite a lot!